overlord/ifacestate/hotplug.go

```diff
@@ -86,14 +106,210 @@ func defaultDeviceKey(devinfo *hotplug.HotplugDeviceInfo, keyVersion int) (strin
+		key, err := deviceKey(defaultKey, devinfo, iface)
+		if err != nil {
+			logger.Noticef("cannot compute hotplug key for device with path %s: %s", devinfo.DevicePath(), err.Error())
+			continue
+		}
+
+		// ignore device that is already handled by a gadget slot
+		if gadgetSlots, ok := gadgetSlotsByInterface[iface.Name()]; ok {
+			for _, gslot := range gadgetSlots {
+				if pred, ok := iface.(hotplug.HandledByGadgetPredicate); ok {
+					if pred.HandledByGadget(devinfo, gslot) {
+						logger.Debugf("ignoring device with hotplug key %q, interface %s (handled by gadget slot %s)", key, iface.Name(), gslot.Name)
+						continue InterfacesLoop
+					}
+				}
+			}
+		}
+
+		proposedSlot, err := hotplugHandler.HotplugDeviceDetected(devinfo)
+		if err != nil {
+			logger.Noticef("cannot process hotplug event by the rule of interface %q: %s", iface.Name(), err)
+			continue
+		}
+		if proposedSlot == nil {
+			continue
+		}
+		if key == "" {
+			logger.Debugf("no valid hotplug key provided by interface %q, device with path %s ignored", iface.Name(), devinfo.DevicePath())
+			continue
+		}
+		proposedSlot, err = proposedSlot.Clean()
+		if err != nil {
+			logger.Noticef("cannot validate hotplug slot proposed by interface %q: %v", iface.Name(), err.Error())
+			continue
+		}
+		if proposedSlot.Label == "" {
+			si := interfaces.StaticInfoOf(iface)
+			proposedSlot.Label = si.Summary
+		}
+
+		if !hotplugFeature {
+			logger.Noticef("hotplug device add event ignored, enable experimental.hotplug")
+			return
+		}
+
+		logger.Debugf("adding hotplug device with path %s for interface %q, hotplug key %s", devinfo.DevicePath(), iface.Name(), key)
+
+		// enumeratedDeviceKeys is only available when enumerating devices
+		if m.enumeratedDeviceKeys != nil {
+			if m.enumeratedDeviceKeys[iface.Name()] == nil {
+				m.enumeratedDeviceKeys[iface.Name()] = make(map[string]bool)
+			}
+			m.enumeratedDeviceKeys[iface.Name()][key] = true
+		}
+		devPath := devinfo.DevicePath()
+		m.hotplugDevicePaths[devPath] = append(m.hotplugDevicePaths[devPath], deviceData{hotplugKey: key, ifaceName: iface.Name()})
+
+		chg := st.NewChange(fmt.Sprintf("hotplug-add-slot-%s", iface), fmt.Sprintf("Add hotplug slot of interface %s with hotplug key %q", iface.Name(), key))
+		hotplugAdd := st.NewTask("hotplug-add-slot", fmt.Sprintf("Create slot for device with hotplug key %q", key))
+		setHotplugAttrs(hotplugAdd, iface.Name(), key)
+		hotplugAdd.Set("device-info", devinfo)
+		hotplugAdd.Set("proposed-slot", proposedSlot)
+		chg.AddTask(hotplugAdd)
+
+		hotplugConnect := st.NewTask("hotplug-connect", fmt.Sprintf("Recreate connections of interface %s, hotplug key %q", iface.Name(), key))
+		setHotplugAttrs(hotplugConnect, iface.Name(), key)
+		hotplugConnect.WaitFor(hotplugAdd)
+		chg.AddTask(hotplugConnect)
+
+		if err := addHotplugSeqWaitTask(chg, key); err != nil {
+			logger.Noticef(err.Error())
```

Comment:
why is this only logged ?

Reply:
Thanks for catching, refactored the helper slightly as discussed on IRC to allocate sequence early and only if there is no error create tasks and change.